Core/clim-basic/extended-streams/recording.lisp

```diff
@@ -142,45 +142,41 @@ all of FUNCTION-ARGS as APPLY arguments."
-                                slot-names))
-         (key-args (mapcar #'(lambda (slot-spec supplied)
-                               `(,slot-spec nil ,supplied))
-                           slots supplied-vars))
-         (key-arg-alist (mapcar #'cons slot-names supplied-vars)))
+  "Each element of SLOTS is either a symbol naming a slot
+or (SLOT-NAME ARGUMENT-ONLY-P)."
+  (multiple-value-bind (slot-names key-args key-arg-alist)
+      (loop for slot-spec in slots
+            for (name argument-only-p) = (alexandria:ensure-list slot-spec)
+            for suppliedp = (gensym (format nil "~A-~A" name '#:p))
+            unless argument-only-p collect name into slot-names
+            collect `(,name nil ,suppliedp) into key-args
+            collect (cons name suppliedp) into key-arg-alist
+            finally (return (values slot-names key-args key-arg-alist)))
     `(progn
        (defmethod output-record-equal and ((record ,record-type)
                                            (record2 ,record-type))
          (macrolet ((if-supplied ((var &optional (type t)) &body supplied-body)
-                      (declare (ignore var type))
-                      `(progn ,@supplied-body)))
+                      (declare (ignore type))
+                      (when (find var ',slot-names :test #'eq)
+                        `(progn ,@supplied-body))))
            (with-slots ,slot-names record2
              ,@body)))
        (defmethod match-output-records-1 and ((record ,record-type)
                                               &key ,@key-args)
          (macrolet ((if-supplied ((var &optional (type t)) &body supplied-body)
-                      (let ((supplied-var (cdr (assoc var ',key-arg-alist))))
-                        (unless supplied-var
-                          (error "Unknown slot ~S" var))
-                        `(or (null ,supplied-var)
-                             ,@(if (eq type t)
-                                   `((progn ,@supplied-body))
-                                   `((if (typep ,var ',type)
-                                         (progn ,@supplied-body)
-                                         (error 'type-error
-                                                :datum ,var
-                                                :expected-type ',type))))))))
+                      (let ((supplied-var (or (cdr (assoc var ',key-arg-alist))
+                                              (error "Unknown argument ~S" var))))
+                        `(cond ((null ,supplied-var)
+                                t)
+                               ,@(unless (eq type t)
+                                   `(((not (typep ,var ',type))
+                                      (error 'type-error
+                                             :datum ,var
+                                             :expected-type ',type))))
+                               (t
+                                ,@supplied-body)))))
```

Comment:
I find the previous version of this block much more readable, please only change the error string.

Reply:
Opposite for me but I change it back if you want.

Reply:
I would still change `,@(if … '((…)) '((…)))` to `,(if … '(…) '(…))` (backticks replaced with single quote for markdown compatibility) though, unless there is reason not to that I missed.

Reply:
I don't see why not either. you may also pull the error to let with or like you do already.

Reply:
Done.